### PR TITLE
Add support for unbundled events in event collection

### PR DESCRIPTION
**Pull request by vboulineau**
### What does this PR do?

Add support for unbundled events in event collection (matching existing support in Helm chart)

### Motivation

Feature parity required for Autoscaling.

### Describe your test plan

In the CRD it's now possible to set:
```
features:
  eventCollection:
    unbundleEvents: true
```

To enable event unbundling in the APIServer check. When enabled, only a subset of events (see doc) will be collected and sent as unbundled (e.g. one Datadog event for each Kubernetes event).

### Checklist

- [x] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [x] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 1]
```diff
@@ -581,6 +581,23 @@ type EventCollectionFeatureConfig struct {
 	// CollectKubernetesEvents enables Kubernetes event collection.
 	// Default: true
 	CollectKubernetesEvents *bool `json:"collectKubernetesEvents,omitempty"`
+
+	// UnbundleEvents enables collection of Kubernetes events as individual events.
+	// Default: false
+	UnbundleEvents *bool `json:"unbundleEvents,omitempty"`
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:587]: should these parameters have `+optional` ?

File: controllers/datadogagent/feature/eventcollection/feature.go [hunk 2]
```diff
@@ -54,6 +74,12 @@ func (f *eventCollectionFeature) Configure(dda *v2alpha1.DatadogAgent) (reqComp
 		f.serviceAccountName = v2alpha1.GetClusterAgentServiceAccount(dda)
 		f.rbacSuffix = common.ClusterAgentSuffix
 
+		if apiutils.BoolValue(dda.Spec.Features.EventCollection.UnbundleEvents) && len(dda.Spec.Features.EventCollection.CollectedUnbundledEvents) > 0 {
+			f.configMapName = apicommonv1.GetConfName(dda, nil, apicommon.DefaultKubeAPIServerConf)
+			f.unbundleEvents = *dda.Spec.Features.EventCollection.UnbundleEvents
+			f.unbundleEventTypes = dda.Spec.Features.EventCollection.CollectedUnbundledEvents
+		}
```
Comment (celenechang) [inline controllers/datadogagent/feature/eventcollection/feature.go:81]: should we log a warning if `UnbundleEvents` is set to `true` but for some reason `CollectedUnbundledEvents == 0` ?

**Review by celenechang — COMMENTED** [review]
just some nit comments/suggestions

File: apis/datadoghq/v2alpha1/datadogagent_types.go (hunk 1 above)
Comment (vboulineau) [reply apis/datadoghq/v2alpha1/datadogagent_types.go:588]: It's implicit already as nullable type and `omitempty` but added as others are explicitly flagged as well.

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 3]
```diff
@@ -581,6 +581,32 @@ type EventCollectionFeatureConfig struct {
...
+	// CollectedEventTypes defines the list of events to collect when UnbundleEvents is enabled.
+	// Default to:
+	// [
+	// {"kind":"Pod","reasons":["Failed","BackOff","Unhealthy","FailedScheduling","FailedMount","FailedAttachVolume"]},
+	// {"kind":"Node","reasons":["TerminatingEvictedPod","NodeNotReady","Rebooted","HostPortConflict"]},
+	// {"kind":"CronJob","reasons":["SawCompletedJob"]}
+	// ]
+	// +optional
+	// +listType=atomic
+	CollectedEventTypes []EventTypes `json:"collectedUnbundledEvents,omitempty"`
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:599]:
```suggestion
	CollectedEventTypes []EventTypes `json:"collectedEventTypes,omitempty"`
```

File: docs/configuration.v2alpha1.md [hunk 4]
```diff
@@ -94,6 +94,8 @@ spec:
 | features.dogstatsd.unixDomainSocketConfig.path | Path defines the socket path used when enabled. |
 | features.ebpfCheck.enabled | Enables the eBPF check. Default: false |
 | features.eventCollection.collectKubernetesEvents | CollectKubernetesEvents enables Kubernetes event collection. Default: true |
+| features.eventCollection.collectedEventTypes | CollectedEventTypes defines the list of events to collect when UnbundleEvents is enabled. Default to: [ {"kind":"Pod","reasons":["Failed","BackOff","Unhealthy","FailedScheduling","FailedMount","FailedAttachVolume"]}, {"kind":"Node","reasons":["TerminatingEvictedPod","NodeNotReady","Rebooted","HostPortConflict"]}, {"kind":"CronJob","reasons":["SawCompletedJob"]} ] |
```
Comment (drichards-87) [inline docs/configuration.v2alpha1.md:97]:
```suggestion
| features.eventCollection.collectedEventTypes | CollectedEventTypes defines the list of events to collect when UnbundleEvents is enabled. Defaults to: [ {"kind":"Pod","reasons":["Failed","BackOff","Unhealthy","FailedScheduling","FailedMount","FailedAttachVolume"]}, {"kind":"Node","reasons":["TerminatingEvictedPod","NodeNotReady","Rebooted","HostPortConflict"]}, {"kind":"CronJob","reasons":["SawCompletedJob"]} ] |
```

**Review by drichards-87 — APPROVED** [review]
Left a tiny suggestion from Docs and approved the PR.

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 5]
```diff
@@ -581,6 +581,32 @@ type EventCollectionFeatureConfig struct {
...
 	// Default: true
 	CollectKubernetesEvents *bool `json:"collectKubernetesEvents,omitempty"`
+
+	// UnbundleEvents enables collection of Kubernetes events as individual events.
+	// Default: false
+	// +optional
+	UnbundleEvents *bool `json:"unbundleEvents,omitempty"`
+
+	// CollectedEventTypes defines the list of events to collect when UnbundleEvents is enabled.
+	// Default to:
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:591]: related to the suggestion from docs, maybe change to this for consistency:
```suggestion
	// Default:
```

**Review by celenechang — APPROVED** [review]

**Pull request merged**